components/mediation-monitor/mediation-data-publisher/org.wso2.carbon.prometheus.publisher/src/main/java/org/wso2/carbon/prometheus/publisher/publisher/MetricCollector.java

```diff
@@ -0,0 +1,358 @@
+ */
+package org.wso2.carbon.prometheus.publisher.publisher;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.prometheus.publisher.model.PrometheusMetric;
+
+import java.lang.management.ManagementFactory;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.management.Attribute;
+import javax.management.AttributeList;
+import javax.management.InstanceNotFoundException;
+import javax.management.IntrospectionException;
+import javax.management.MBeanAttributeInfo;
+import javax.management.MBeanInfo;
+import javax.management.MBeanServer;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectInstance;
+import javax.management.ObjectName;
+import javax.management.ReflectionException;
+
+class MetricCollector {
+
+    private static final Log LOGGER = LogFactory.getLog(MetricCollector.class);
+
+    private static final String SYNAPSE_DOMAIN_NAME = "org.apache.synapse:*";
+    private static final String METRIC_TYPE_UNTYPED = "untyped";
+    private static final String SEPARATOR = "_";
+    private static final Pattern PROPERTY_PATTERN = Pattern.compile(
+            "([^,=:*?]+)" + // Name - non-empty, anything but comma, equals, colon, star, or question mark
+                    "=" +  // Equals
+                    "(" + // Either
+                    "\"" + // Quoted
+                    "(?:" + // A possibly empty sequence of
+                    "[^\\\\\"]*" + // Greedily match anything but backslash or quote
+                    "(?:\\\\.)?" + // Greedily see if we can match an escaped sequence
+                    ")*" +
+                    "\"" +
+                    "|" + // Or
+                    "[^,=:\"]*" + // Unquoted - can be empty, anything but comma, equals, colon, or quote
+                    ")");
+
+    private static volatile MetricCollector metricCollector;
+
+//    private MetricCollector() {
+//
+//    }
+
+    /**
+     * @return Single instance of MetricCollector
+     */
+    static MetricCollector getInstance() {
+
+        MetricCollector collector = metricCollector;
+        if (collector == null) {
+            synchronized (MetricCollector.class) {
+                collector = metricCollector;
+                if (collector == null) {
+                    metricCollector = collector = new MetricCollector();
+                }
+            }
+        }
+        return collector;
+    }
+
+    /**
+     * Collects metric data from JMX MBeans
+     *
+     * @return Metric data in Prometheus format
+     */
+    List<String> collect() {
+
+        List<PrometheusMetric> metricList = new ArrayList<>();
```

Comment:
There can be race condition issues with this field since the class is a singleton. Shall we check whether we can make it not a singleton class?